features/nfc/source/nfc/NFCEEPROM.cpp

```diff
@@ -0,0 +1,409 @@
+            _current_op = nfc_eeprom_write_end_session;
+            _operation_result = NFC_OK;
+            _driver->end_session();
+            break;
+        case nfc_eeprom_erase_write_size:
+            if (!success) {
+                handle_error(NFC_ERR_CONTROLLER);
+                return;
+            }
+
+            // End session
+            _current_op = nfc_eeprom_erase_end_session;
+            _operation_result = NFC_OK;
+            _driver->end_session();
+            break;
+        default:
+            // Should not happen, state machine is broken or driver is doing something wrong
+            handle_error(NFC_ERR_UNKNOWN);
+            return;
+    }
+}
+
+void NFCEEPROM::on_size_read(bool success, size_t size)
+{
+    switch (_current_op) {
+        case nfc_eeprom_read_read_size: {
+            if (!success) {
+                handle_error(NFC_ERR_CONTROLLER);
+                return;
+            }
+
+            // Reset NDEF message buffer builder
+            ac_buffer_builder_t *buffer_builder = ndef_msg_buffer_builder(ndef_message());
+            ac_buffer_builder_reset(buffer_builder);
+
+            // Check that we have a big enough buffer to read the message
+            if (size > ac_buffer_builder_writable(buffer_builder)) {
+                // Not enough space, close session
+                _current_op = nfc_eeprom_read_end_session;
+                _operation_result = NFC_ERR_BUFFER_TOO_SMALL;
+                _driver->end_session();
+                return;
+            }
+
+            // Save size and reset address
+            _eeprom_address = 0;
+            _ndef_buffer_read_sz = size;
+
+            // Start reading bytes
+            _current_op = nfc_eeprom_read_read_bytes;
+            continue_read();
+            break;
+        }
+        default:
+            // Should not happen, state machine is broken or driver is doing something wrong
+            handle_error(NFC_ERR_UNKNOWN);
+            return;
+    }
+}
+
+void NFCEEPROM::on_bytes_erased(size_t count)
+{
+    switch (_current_op) {
+        case nfc_eeprom_erase_erase_bytes:
+            if (count == 0) {
+                handle_error(NFC_ERR_CONTROLLER);
+                return;
+            }
+
+            // Update address
+            _eeprom_address += count;
+
+            // Continue erasing
+            continue_erase();
+            break;
+        default:
+            // Should not happen, state machine is broken or driver is doing something wrong
+            handle_error(NFC_ERR_UNKNOWN);
+            return;
+    }
+}
+
+void NFCEEPROM::continue_write()
+{
+    if (ac_buffer_reader_readable(&_ndef_buffer_reader) > 0) {
+        // Continue writing
+        _driver->write_bytes(_eeprom_address, ac_buffer_reader_current_buffer_pointer(&_ndef_buffer_reader), ac_buffer_reader_current_buffer_length(&_ndef_buffer_reader));
+    } else {
+        // Now update size
+        _current_op = nfc_eeprom_write_write_size;
```

Comment:
size needs to be set first, otherwise bytes written beyond size will be truncated

Reply:
The size is size of memory, not size of ndef message. You have a field for size in the ndef message for that.

Reply:
...also this is how the driver and tests are implemented and documented: you can only read and write memory to within the currently set size.